python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -194,6 +196,45 @@ def start_time_for_partition_key(self, partition_key: str) -> datetime:
         # we make the assumption that the parsed partition key is <= the start datetime
         return next(iter(self._iterate_time_windows(partition_key_dt))).start
 
+    def time_windows_for_partition_keys(
+        self, partition_keys: Sequence[str], _valid_time_windows: Optional[List[TimeWindow]] = None
+    ) -> List[TimeWindow]:
+        """
+        Accepts an optional `_valid_time_windows` argument to make creating multipartitions subsets
+        more performant. This avoids refetching the list of valid time windows for each time window
+        partitions subset.
+        """
+        if len(partition_keys) == 0:
+            return []
+
+        partition_keys = list(set(partition_keys))
+        sorted_key_datetimes = sorted(
+            set(
+                [
+                    pendulum.instance(datetime.strptime(partition_key, self.fmt), tz=self.timezone)
+                    for partition_key in partition_keys
+                ]
+            ),
+            key=lambda dt: dt.timestamp(),
+        )
+        curr_key_i = 0
+        partition_key_time_windows = []
+        for time_window in (
+            self.iterate_valid_time_windows(self.start)
+            if not _valid_time_windows
+            else _valid_time_windows
+        ):
+            if time_window.start.timestamp() == sorted_key_datetimes[curr_key_i].timestamp():
```

Comment:
I don't think doing this comparison will always work, because sometimes a partition named "2020-01-01" will correspond to a window timestamp of that day at 7 pm.